src/lib.rs

```diff
@@ -306,54 +126,202 @@ fn infallible<T>(result: Result<T, CollectionAllocErr>) -> T {
+            self.as_mut_ptr_heap()
+        } else {
+            self.as_mut_ptr_inline()
+        };
+        let len = len.value(Self::is_zst());
+
+        let new_layout =
+            Layout::array::<T>(new_capacity).map_err(|_| CollectionAllocErr::CapacityOverflow)?;
+        if new_layout.size() > isize::MAX as usize {
+            return Err(CollectionAllocErr::CapacityOverflow);
+        }
+
+        if len == 0 || !was_on_heap {
+            // get a fresh allocation
+
+            // layout has non zero size
+            let new_ptr = alloc(new_layout) as *mut T;
+            if new_ptr.is_null() {
+                Err(CollectionAllocErr::AllocErr { layout: new_layout })
+            } else {
+                copy_nonoverlapping(ptr, new_ptr, len);
+                *self = Self::new_heap(new_ptr, new_capacity);
+                Ok(())
+            }
+        } else {
+            // use realloc
+
+            // this can't overflow since we already constructed an equivalent layout during
+            // the previous allocation
+            let old_layout =
+                Layout::from_size_align_unchecked(self.heap.1 * size_of::<T>(), align_of::<T>());
+
+            // SAFETY: ptr was allocated with this allocator
+            // old_layout is the same as the layout used to allocate the previous memory block
+            // new_layout.size() is greater than zero
+            // does not overflow when rounded up to alignment. since it was constructed
+            // with Layout::array
+            let new_ptr = realloc(ptr as *mut u8, old_layout, new_layout.size()) as *mut T;
+            if new_ptr.is_null() {
+                Err(CollectionAllocErr::AllocErr { layout: new_layout })
+            } else {
+                *self = Self::new_heap(new_ptr, new_capacity);
+                Ok(())
+            }
+        }
+    }
+}
+
+/// Vec guarantees that its length is always less than isize::MAX in *bytes*.
+///
+/// For a non ZST, this means that the length is less than isize::MAX objects, which implies we
+/// have at least one free bit we can use. We use the least significant bit for the tag. And store
+/// the length in the `usize::BITS - 1` most significant bits.
+///
+/// For a ZST, we never use the heap, so we just store the length directly.
+#[repr(transparent)]
+#[derive(Clone, Copy)]
+struct TaggedLen(usize);
+
+impl TaggedLen {
+    #[inline]
+    pub const fn new(len: usize, on_heap: bool, is_zst: bool) -> Self {
+        if is_zst {
+            debug_assert!(!on_heap);
+            TaggedLen(len)
+        } else {
+            debug_assert!(len < isize::MAX as usize);
+            TaggedLen((len << 1) | on_heap as usize)
+        }
+    }
+
+    #[inline]
+    #[must_use]
+    pub const fn on_heap(self, is_zst: bool) -> bool {
+        if is_zst {
+            false
+        } else {
+            (self.0 & 1_usize) == 1
+        }
+    }
+
+    #[inline]
+    pub const fn value(self, is_zst: bool) -> usize {
```

Comment:
What do you thank of making `TaggedLen` be generic (`TaggedLen<T>`), so we wouldn't need to pass `is_zst` arguments to its methods?

Reply:
there's the issue of the variance of `TaggedLen` with respect to `T`. it has to be covariant so that the vector is also covariant, but i don't feel like that option makes sense to me since it doesn't really contain a `T` (maybe have it contain a `PhantomData<fn() -> T>`?). ideally i'd like to have `TaggedLen<const IS_ZST: bool>(...)` but we run into current limitations of const generics when doing that.

i'd personally keep it as is but the decision is up to you

Reply:
Okay, that makes sense.

Reply:
Sorry again for the radio silence.  My plan is to merge this as-is, effectively making this branch the new mainline, and open issues for porting the remaining changes and any other missing pieces.  The current `master` branch will become a `v1` branch, for bug fixes only.  I'm not sure when I'll have time to get to this, but it's still on my to-do list.

Reply:
no problem! let me know if i can be of any help